src/_common/content/content-editor/controls/mention/autocomplete/controls.ts

```diff
@@ -0,0 +1,223 @@
+import { EditorView } from 'prosemirror-view';
+import Vue from 'vue';
+import { Component, Prop, Watch } from 'vue-property-decorator';
+import { Api } from '../../../../../api/api.service';
+import AppLoading from '../../../../../loading/loading.vue';
+import { Screen } from '../../../../../screen/screen-service';
+import AppUserAvatarImg from '../../../../../user/user-avatar/img/img.vue';
+import { User } from '../../../../../user/user.model';
+import AppUserVerifiedTick from '../../../../../user/verified-tick/verified-tick.vue';
+import { ContentEditorService } from '../../../content-editor.service';
+import { ContentEditorSchema } from '../../../schemas/content-editor-schema';
+import ContentEditorMentionCache from '../cache.service';
+
+@Component({
+	components: {
+		AppLoading,
+		AppUserAvatarImg,
+		AppUserVerifiedTick,
+	},
+})
+export default class AppContentEditorControlsMentionAutocompleteControls extends Vue {
+	@Prop(Object)
+	view!: EditorView<ContentEditorSchema>;
+	@Prop(Number)
+	stateCounter!: number;
+	@Prop(Number)
+	canShow!: number;
+
+	visible = false;
+	top = 'auto';
+	bottom = 'auto';
+	left = 'auto';
+
+	query = ''; // Currently active suggestion query
+	selectedIndex = 0;
+	isListening = false; // If we are listening to the document keydown event, to be able to unbind it later.
+	isInverted = false; // If the list is inverted due to screen size constraints. It shows above the control instead if below.
+	remoteSuggestionTimeout?: NodeJS.Timer; // Timeout between requests to search backend
+	isLoading = false; // Loading more users from backend
+	users: User[] = [];
+
+	readonly Screen = Screen;
+
+	$refs!: {
+		container: HTMLElement;
+		list: HTMLDivElement;
+	};
+
+	get displayUsers() {
+		if (this.isInverted) {
+			return this.users.slice().reverse();
+		}
+		return this.users;
+	}
+
+	get showControl() {
+		return this.visible && (this.isLoading || this.users.length > 0);
+	}
+
+	mounted() {
+		this.update();
+		document.addEventListener('keydown', this.onKeyDown);
+		this.isListening = true;
+	}
+
+	destroyed() {
+		if (this.isListening) {
+			document.removeEventListener('keydown', this.onKeyDown);
+			this.isListening = false;
+		}
+
+		if (this.remoteSuggestionTimeout) {
+			clearTimeout(this.remoteSuggestionTimeout);
+			this.remoteSuggestionTimeout = undefined;
+		}
+	}
+
+	@Watch('canShow')
+	@Watch('stateCounter')
+	private update() {
+		if (this.canShow && this.view instanceof EditorView) {
+			const state = this.view.state;
+			const node = ContentEditorService.getSelectedNode(this.view.state);
+
+			if (node !== null && node.isText) {
+				// Get all text from the beginning of the doc to the end of the selection.
+				// Then from the end of that slice, get the mention text.
+				const slice = state.doc.slice(0, state.selection.from);
+				const text = ContentEditorService.getFragmentText(slice.content);
+				const matches = /@([\w_-]+)$/i.exec(text);
```

Comment:
The following condition is not to check the length of the result, but the length of the regex match result. The first one is the entire match, the second one the capture. So it has to at least be two items so I can make sure the capture worked.

Reply:
oh yeah, sorry. didnt notice its not an index check haha.
still, the minimum length should be in the regex and you didnt address the reusing bit. later we might come to this and forget this regex is duplicated elsewhere in the code.

Reply:
I did reuse it.

Reply:
Oh, and it should start showing from the first character after the @. Of course mentions cannot be less than 2 characters because that's the minimum for usernames, but I think it still makes sense to show suggestions. When selecting the suggestion, it would still be the correct length.

Reply:
the crap? github actually did not show me the new changes until after i hard refreshed.

Reply:
allgud